libdevcore/LevelDBSnapManager.cpp

```diff
@@ -0,0 +1,115 @@
+// this function returns the size of oldSnaps after cleanup
+uint64_t LevelDBSnapManager::garbageCollectUnusedOldSnaps(
+    std::unique_ptr< leveldb::DB >& _db, uint64_t _dbReopenId, uint64_t _maxSnapLifetimeMs ) {
+    std::vector< std::shared_ptr< LevelDBSnap > > unusedOldSnaps;
+
+    uint64_t mapSizeAfterCleanup;
+
+    auto currentTimeMs = LevelDB::getCurrentTimeMs();
+
+    {
+        // find all old snaps and remove them from the map
+        // this needs to be done write lock so the map is not concurrently modified
+        std::unique_lock< std::shared_mutex > snapLock( m_snapMutex );
+        for ( auto it = oldSnaps.begin(); it != oldSnaps.end(); ) {
+            // a snap is unused if no one using this snap anymore except the map itself
+            if ( it->second.use_count() == 1 ||
+                 it->second->getCreationTimeMs() + _maxSnapLifetimeMs <= currentTimeMs ) {
+                // erase this snap from the map
+                // note that push back needs to happen before erase since erase will
+                // destroy the snap object if there ar no more references
+                unusedOldSnaps.push_back( it->second );
+                it = oldSnaps.erase( it );
+            } else {
+                ++it;  // Only increment if not erasing
+            }
+        }
+        mapSizeAfterCleanup = oldSnaps.size();
+    }
+
+    // now we removed unused snaps from the map. Close them
+
+    for ( auto&& snap : unusedOldSnaps ) {
+        LDB_CHECK( snap );
+        snap->close( _db, _dbReopenId );
+    }
+    return mapSizeAfterCleanup;
+}
+
+
+// this will be called from EVM processing thread just after the block is processed
+void LevelDBSnapManager::addSnapForBlock(
+    uint64_t _blockNumber, std::unique_ptr< leveldb::DB >& _db, uint64_t _dbInstanceId ) {
+    createNewSnap( _blockNumber, _db, _dbInstanceId );
+
+    // we garbage-collect  unused old snaps that no-one used or that exist for more that max
+    // lifetime we give for eth_calls to complete
+    garbageCollectUnusedOldSnaps( _db, _dbInstanceId, OLD_SNAP_LIFETIME_MS );
+}
+
+
+// this will create new last block snap and move previous last block snap into old snaps map
+void LevelDBSnapManager::createNewSnap(
+    uint64_t _blockId, std::unique_ptr< leveldb::DB >& _db, uint64_t _dbInstanceId ) {
+    // hold the write lock during the update so snap manager does not return inconsistent value
+    // snap creation in LevelDB should happen really fast
+    std::unique_lock< std::shared_mutex > lock( m_snapMutex );
+
+    LDB_CHECK( _db );
+    auto newSnapHandle = _db->GetSnapshot();
+    LDB_CHECK( newSnapHandle );
+
+    auto newSnap = std::make_shared< LevelDBSnap >( _blockId, newSnapHandle, _dbInstanceId );
+    LDB_CHECK( newSnap );
+
+    auto oldSnap = m_lastBlockSnap;
+    m_lastBlockSnap = newSnap;
+    if ( oldSnap ) {
+        oldSnaps.emplace( oldSnap->getInstanceId(), oldSnap );
```

Comment:
shouldn't it be `oldSnaps.emplace( oldSnap->getBlockId(), oldSnap, oldSnap->getInstanceId());` ?